### PR TITLE
feat(site): render LaTeX math via KaTeX + convert plain-text formulas

### LINKED ISSUE #158: Feature Request: Convert plain text math formulas to LaTeX rendering
### 📋 Overview
Currently, some mathematical formulas in the project documentation/pages are written in plain text or standard Markdown, which significantly impacts readability. I suggest converting these formulas into LaTeX format so they can be properly rendered using standard tools like MathJax or KaTeX.

### 🔄 Formula Mapping

Here are some examples:

| Context | Current Plain Text Formula | LaTeX Formula | The rendering result of LaTeX |
| :--- | :--- | :--- |:---|
| Quadratic formula | x = (-b +/- sqrt(b^2 - 4ac)) / 2a | `$x = \frac{-b \pm \sqrt{b^2 - 4ac}}{2a}$` |$x = \frac{-b \pm \sqrt{b^2 - 4ac}}{2a}$|
| The formal definition | f'(x) = limit (h -> 0) [f(x + h) - f(x)] / h | `$f'(x) = \lim\limits_{h \to 0} \frac{f(x + h) - f(x) }{h}$` |$f'(x) = \lim\limits_{h \to 0} \frac{f(x + h) - f(x) }{h}$|
| The definition of gradient | grad f = [ df/dx, df/dy, df/dz ] | `$\text{grad } f = \left[ \frac{\partial f}{\partial x}, \frac{\partial f}{\partial y}, \frac{\partial f}{\partial z}\right]$` |$\text{grad } f = \left[ \frac{\partial f}{\partial x}, \frac{\partial f}{\partial y}, \frac{\partial f}{\partial z}\right]$|

### 🛠️ Suggested Implementation
- For inline formulas, use single dollar signs: `$ formula $`
- For standalone block formulas, use double dollar signs: `$$ formula $$`

### PR DESCRIPTION
**Pull request by Jayantkhandebharad**
Closes #158.

Math formulas across the curriculum were written as ASCII art in untagged code fences or as plain text inline, which reads poorly. The blocker was that **the site had no math renderer at all** — so simply writing `$...$` would have shown literal dollar signs. This PR adds the renderer and converts the formulas.

## 1. Render LaTeX with KaTeX (`site/lesson.html`)

- Pinned **KaTeX 0.17.0** via CDN with **SRI integrity hashes**.
- Extended the client-side markdown renderer to support inline `$...$` and block `$$...$$` (single- and multi-line, including `\begin{aligned}` and `bmatrix`).
- Inline math uses the **Pandoc rule** (opening `$` not followed by a space; closing `$` not preceded by a space nor followed by a digit) so prose dollar amounts that are common in this curriculum — `$50,000`, `$0.25/$2.50` — are never mistaken for math. Verified **0 false positives across all 503 docs**.
- Inline code and math are stashed behind sentinels before the escape/emphasis pass, so a `$` inside `` `code` `` and a `*` inside a formula stay verbatim. Invalid or unloaded TeX degrades to the raw source rather than blanking the lesson.

## 2. Convert plain-text formulas (54 lessons, one commit each)

Genuine formulas were converted to `$...$` / `$$...$$`; the math is unchanged, only the notation. Architecture diagrams, terminal sessions, code, mermaid blocks, and shape-annotation sketches were deliberately left as monospace.

The three examples from the issue now render, e.g.:

- `$x = \frac{-b \pm \sqrt{b^2 - 4ac}}{2a}$`
- `$f'(x) = \lim_{h \to 0} \frac{f(x+h) - f(x)}{h}$`
- `$\nabla f = \left[ \frac{\partial f}{\partial x}, \frac{\partial f}{\partial y}, \frac{\partial f}{\partial z} \right]$`

## Verification

- **1,742 math spans** (1,448 inline + 294 block) compile-checked with KaTeX `throwOnError` → **0 failures**; **0** literal pipes inside table-cell math.
- Inline/block parsing covered by unit tests of the real parser (money-safety, `$`-in-code, `*`-in-math, edge cases).
- Live browser render confirmed for the issue examples and for converted content (aligned derivations, `bmatrix`, inline-in-prose, table-cell math).
- `python3 scripts/audit_lessons.py` → **0 issues** across 503 lessons. No new broken links introduced.

`site/data.js` is left for CI to regenerate on merge, per `AGENTS.md`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: phases/01-math-foundations/07-bayes-theorem/docs/en.md [hunk 1]
```diff
@@ -410,22 +411,22 @@ Setup: you are testing two button colors. Variant A (blue) and variant B (green)
...
+4. **Decision.** Compute $P(B > A)$ -- the probability that B's true conversion rate is higher than A's.
 
-Computing P(B > A) analytically is hard. But Monte Carlo makes it trivial:
+Computing $P(B > A)$ analytically is hard. But Monte Carlo makes it trivial:
 
 ```
 1. Draw 100,000 samples from Beta(51, 951)  -> samples_A
 2. Draw 100,000 samples from Beta(66, 936)  -> samples_B
 3. P(B > A) = fraction of samples where B > A
 ```
```
[the remark this answers is not in the document]
Comment (Jayantkhandebharad) [reply phases/01-math-foundations/07-bayes-theorem/docs/en.md]: Fixed in 62167a0 — tagged the fence as ```text.

File: phases/01-math-foundations/13-numerical-stability/docs/en.md [hunk 2]
```diff
@@ -568,13 +570,13 @@ These stable implementations reappear in Phase 3 when building the training loop
 
 ## Exercises
 
-1. **Catastrophic cancellation.** Compute the variance of [1000000.0, 1000001.0, 1000002.0] using the naive formula `E[x^2] - E[x]^2` in float32. Then compute it using Welford's online algorithm. Compare the errors against the true variance (0.6667).
+1. **Catastrophic cancellation.** Compute the variance of [1000000.0, 1000001.0, 1000002.0] using the naive formula $E[x^2] - E[x]^2$ in float32. Then compute it using Welford's online algorithm. Compare the errors against the true variance (0.6667).
 
 2. **Precision hunt.** Find the smallest positive float32 value `x` such that `1.0 + x == 1.0` in Python. This is the machine epsilon. Verify it matches `numpy.finfo(numpy.float32).eps`.
```
[the remark this answers is not in the document]
Comment (Jayantkhandebharad) [reply phases/01-math-foundations/13-numerical-stability/docs/en.md]: Good catch — fixed in 7e3a6ee. The exercise now asks for the smallest `x` with `1.0 + x > 1.0` (the gap above 1.0, i.e. machine epsilon), which matches `numpy.finfo(numpy.float32).eps` = 1.1920929e-07.

File: phases/01-math-foundations/14-norms-and-distances/docs/en.md [hunk 3]
```diff
@@ -28,15 +28,15 @@ This lesson builds every major distance function from scratch, shows you when ea
 
 ### Norms: measuring vector magnitude
 
-A norm measures the "size" of a vector. Every distance function between two vectors can be written as the norm of their difference: d(a, b) = ||a - b||. So understanding norms is understanding distances.
+A norm measures the "size" of a vector. Every distance function between two vectors can be written as the norm of their difference: $d(a, b) = \|a - b\|$. So understanding norms is understanding distances.
```
[the remark this answers is not in the document]
Comment (Jayantkhandebharad) [reply phases/01-math-foundations/14-norms-and-distances/docs/en.md]: Fixed in 0e924ee — qualified to "Many vector distances…" since not every metric is norm-induced (e.g. cosine distance, covered later in this lesson).

File: phases/01-math-foundations/15-statistics-for-ml/docs/en.md [hunk 4]
```diff
@@ -174,28 +167,24 @@ Example:
...
+$$
+\bar{x} \pm z \cdot \frac{s}{\sqrt{n}}
+$$
 
-Interpretation: if you repeated this experiment many times, 95% of the
-computed intervals would contain the true mean. It does NOT mean there
-is a 95% probability the true mean is in this specific interval.
-```
+where $z = 1.96$ for 95% confidence.
+
```
[the remark this answers is not in the document]
Comment (Jayantkhandebharad) [reply phases/01-math-foundations/15-statistics-for-ml/docs/en.md:186]: Fixed in 64754c8. Kept the simple $z = 1.96$ form for the intro but added a caveat: for small $n$ with unknown population variance, use the t-critical value $t_{n-1,\,0.975}$.

File: phases/01-math-foundations/16-sampling-methods/docs/en.md [hunk 5]
```diff
@@ -44,170 +44,175 @@ The core challenge: you can only sample directly from simple distributions (unif
...
+&\text{Proposal distribution: } q(x) \quad (\text{can sample from}) \\
+&\text{Bound: } M \text{ such that } p(x) \leq M \, q(x) \text{ for all } x \\
+&\text{Algorithm:} \\
+&\quad 1.\ \text{Sample } x \sim q(x) \\
+&\quad 2.\ \text{Sample } u \sim \text{Uniform}(0, 1) \\
+&\quad 3.\ \text{If } u < p(x) / (M \, q(x)), \text{ accept } x \\
+&\quad 4.\ \text{Otherwise, reject and go to step 1} \\
+&\text{Acceptance rate} = 1/M
+\end{aligned}
+$$
```
[the remark this answers is not in the document]
Comment (Jayantkhandebharad) [reply phases/01-math-foundations/16-sampling-methods/docs/en.md:123]: Fixed in e649c7c — appended `\quad \text{(when } p \text{ is normalized)}` to the acceptance-rate line.

File: phases/01-math-foundations/17-linear-systems/docs/en.md [hunk 6]
```diff
@@ -323,11 +322,11 @@ small change in x                    huge change in x
...
 Rules of thumb:
-- kappa < 100: safe, solution is accurate.
-- kappa ~ 10^k: you lose about k digits of precision from your floating-point arithmetic.
-- kappa ~ 10^16 (for float64): the solution is meaningless. The matrix is effectively singular.
+- $\kappa < 100$: safe, solution is accurate.
+- $\kappa \sim 10^k$: you lose about $k$ digits of precision from your floating-point arithmetic.
+- $\kappa \sim 10^{16}$ (for float64): the solution is meaningless. The matrix is effectively singular.
 
-In ML, ill-conditioning happens when features are nearly collinear. Regularization (adding lambda * I) improves the condition number from sigma_max / sigma_min to (sigma_max + lambda) / (sigma_min + lambda).
+In ML, ill-conditioning happens when features are nearly collinear. Regularization (adding $\lambda I$) improves the condition number from $\sigma_{\max} / \sigma_{\min}$ to $(\sigma_{\max} + \lambda) / (\sigma_{\min} + \lambda)$.
```
[the remark this answers is not in the document]
Comment (Jayantkhandebharad) [reply phases/01-math-foundations/17-linear-systems/docs/en.md]: Fixed in e0ad0bd. The normal-equations baseline is also squared, so it now reads $\sigma_{\max}^2/\sigma_{\min}^2 \to (\sigma_{\max}^2+\lambda)/(\sigma_{\min}^2+\lambda)$.

File: phases/01-math-foundations/20-fourier-transform/docs/en.md [hunk 7]
```diff
@@ -30,65 +30,62 @@ This matters for ML because frequency-domain thinking appears everywhere. Convol
...
-```
+$$
+X[0] = \sum_{n=0}^{N-1} x[n] \cdot e^0 = \text{sum of all samples}
+$$
 
-**X[k] for 1 <= k <= N/2: positive frequencies.** X[k] represents frequency k cycles per N samples. Higher k means higher frequency (faster oscillation).
+**$X[k]$ for $1 \leq k \leq N/2$: positive frequencies.** $X[k]$ represents frequency $k$ cycles per $N$ samples. Higher $k$ means higher frequency (faster oscillation).
 
-**X[N/2]: the Nyquist frequency.** The highest frequency you can represent with N samples. Above this, you get aliasing -- high frequencies masquerading as low ones.
+**$X[N/2]$: the Nyquist frequency.** The highest frequency you can represent with $N$ samples. Above this, you get aliasing -- high frequencies masquerading as low ones.
```
[the remark this answers is not in the document]
Comment (Jayantkhandebharad) [reply phases/01-math-foundations/20-fourier-transform/docs/en.md:52]: Fixed in 83da9eb — tagged the STFT-procedure and aliasing-example fences as ```text. (The walk-positions block referenced at 45–52 is already a $$ math block.)

**Comment by Jayantkhandebharad** [comment]
## Review feedback addressed

Thanks for the thorough pass. I verified each finding against the actual code and KaTeX before acting. Summary below; all changed math re-validated (KaTeX `throwOnError` compile-check of 643 spans in the touched files → 0 failures; `audit_lessons.py` → 0 issues).

### Fixed

**In-diff (replied inline):** norm-induced distance claim, machine-epsilon exercise, t-interval caveat, rejection-sampling acceptance rate, regularized condition number (squared spectrum), and the untyped fences in bayes/fourier.

**Outside-diff:**
- `11-singular-value-decomposition` — dropped the misleading `image = np.random.seed(42)` assignment (`92af52f`).
- `16-sampling-methods` — `sample_exponential_inverse_cdf` now uses `-math.log1p(-u)` (well-defined at `u = 0`), and `temperature_sample` returns the argmax for `T = 0` instead of dividing by zero, matching the documented "T = 0.0: greedy decoding" behavior (`e649c7c`).
- `17-linear-systems` — corrected the least-squares example to `x = [1.5, 1.6]` (the normal equations `[[4,10],[10,30]] x = [22,63]` solve to exactly that) (`e0ad0bd`).
- `22-stochastic-processes` — varied the three consecutive "At …" sentence openings (`e923e8c`).

### Not changed (verified non-issues)

- **`11-llm-engineering/13-production-app` "critical KaTeX failures" — both are false positives.**
  - The Scaling-table currency (`$50/month`, `$500/month`, `$5K/month`, `$50K+/month`) is **not** matched by the inline regex: every `$` is immediately followed by a digit, so none qualifies as a *closing* delimiter under the Pandoc rule (`\$(?!\d)` on the closer). I re-ran the regex over that exact row → 0 matches, consistent with the 0 false positives this rule produces across all 503 docs.
  - `$50{,}000 \times (1 - 0.35)$` renders fine — `{,}` is a braced group, the standard idiom for a tight thousands-separator in math mode, and it's in the compile-checked span set (`katex.renderToString(..., {throwOnError:true})` passes). No change needed.

- **YAML frontmatter (`stripFrontmatter` in `site/lesson.html`; migrating `docs/en.md` to a `---` block).** This repo's lesson contract (`AGENTS.md`) uses inline `**Type:**` / `**Languages:**` / `**Prerequisites:**` / `**Time:**` metadata, not YAML frontmatter — the renderer special-cases those `**key:** value` lines (`lesson-meta-tag`). Introducing `---` frontmatter would render as an `<hr>` plus visible metadata and diverge from all 503 docs, so I've kept the existing convention. Happy to revisit if the project decides to adopt YAML frontmatter repo-wide (that'd be its own change).

**Comment by Jayantkhandebharad** [comment]
@rohitg00 can you take a look at this one!
